qcodes/utils/dataset/doNd.py

```diff
@@ -495,6 +496,242 @@ def do2d(
+            flat_setpoints = np.vstack(flat_setpoint_grids).T
+        else:
+            flat_setpoints = [()]
+        return flat_setpoints
+
+    def _parse_dond_arguments(*params: Union[_AbstractSweep, ParamMeasT]
+                              ) -> Tuple[
+        List[_BaseParameter],
+        Dict[_BaseParameter, Dict[str, float]],
+        List[ParamMeasT]
+    ]:
+        """
+        Parse supplied arguments for a dond scan.
+        """
+        params_set: List[_BaseParameter] = []
+        params_delay: Dict[_BaseParameter, Dict[str, float]] = {}
+        params_meas: List[ParamMeasT] = []
+        for param in params:
+            if isinstance(param, _AbstractSweep):
+                params_set.append(param._param)
+                params_delay[param._param] = {}
+                params_delay[param._param]['delay'] = param._delay
+            elif isinstance(param, _BaseParameter):
+                params_meas.append(param)
+            else:
+                raise ValueError(f'Cannot proceed: {param} is not either a'
+                                 ' sweep instance or measurement parameter.'
+                                 ' Please check the supplied `dond`'
+                                 ' arguments.')
+        return params_set, params_delay, params_meas
+
+    params_set, params_delay, params_meas = _parse_dond_arguments(*params)
+    nested_setpoints = _make_nested_setpoints(*params)
+
+    all_setpoint_params = tuple(params_set) + tuple(
+            s for s in additional_setpoints)
+
+    measured_parameters = tuple(param for param in params_meas
+                                if isinstance(param, _BaseParameter))
+
+    num_points_params_set: List[int] = []
+    for par in params:
+        if isinstance(par, _AbstractSweep):
+            num_points_params_set.append(int(par._num_points))
+
+    try:
+        loop_shape = tuple(
+            1 for _ in additional_setpoints
+        ) + tuple(num_points_params_set)
+        shapes: Shapes = detect_shape_of_measurement(
+            measured_parameters,
+            loop_shape
+        )
+    except TypeError:
+        LOG.exception(
+            f"Could not detect shape of {measured_parameters} "
+            f"falling back to unknown shape.")
+        shapes = None
+
+    _register_parameters(meas, all_setpoint_params)
+    _register_parameters(
+        meas, params_meas, setpoints=all_setpoint_params, shapes=shapes
+    )
+    _set_write_period(meas, write_period)
+
+    with _catch_keyboard_interrupts() as interrupted, meas.run() as datasaver:
+        additional_setpoints_data = _process_params_meas(additional_setpoints)
+        for setpoints in tqdm(nested_setpoints, disable=not show_progress):
+            param_set_list = []
+            param_value_pairs = zip(params_set[::-1], setpoints[::-1])
+            for setpoint_param, setpoint in param_value_pairs:
+                setpoint_param.post_delay = params_delay[setpoint_param]['delay']
```

Comment:
It should not be needed to set post_delay more than once. E.g. this can be moved outside the for loop. 

We should also ensure that this is reset after the loop. (have a look at the other do1d, do2d functions)